srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,544 @@
+    elements[idx].classList.add(newClass);
+  }
+
+  for (const [breakName, bk] of breaks) {
+    if (bk.width.length > nKids) {
+      const msg = `Truncating number of widths at '${breakName}' breakpoint to match number of elements.`;
+      console.warn(msg, { widths: bk.width.length, elements: nKids });
+    }
+
+    // Find widths, before, after for each element, recycled from the spec
+    const widths = recycleToLength(bk.width, nKids);
+    const before = recycleToLength(bk.before, nKids);
+    const after = recycleToLength(bk.after, nKids);
+
+    let cursor = 0;
+    const updateCursor = (incr: number, isEmpty = false) => {
+      cursor = Math.abs(cursor);
+      let newCursor = cursor + incr;
+      if (newCursor == unitsRow) {
+        // we've reached the final column, allow for a natural break
+        newCursor = 0;
+      }
+      if (newCursor > unitsRow) {
+        // this row is full, empty columns are allowed to break to the next row
+        // which we signal with cursor < 0
+        newCursor = isEmpty ? -1 * (newCursor % unitsRow) : incr;
+      }
+      // console.log(`cursor: ${cursor} -> ${newCursor} (+${incr}${isEmpty ? " (empty)" : ""})`);
+      cursor = newCursor;
+    };
+
+    for (let idx = 0; idx < nKids; idx++) {
+      let addStartClass = false;
+      const unitsMoveAhead = before[idx] + (idx > 0 ? after[idx - 1] : 0);
+      const unitsThisItem = Math.min(widths[idx], unitsRow);
+
+      let unitsRowRemaining = unitsRow - cursor;
+
+      if (unitsMoveAhead > 0) {
+        updateCursor(unitsMoveAhead, true);
+        if (cursor < 0) {
+          cursor = Math.abs(cursor);
+          // adding empty cols caused a row wrap, so we need a start class if
+          // 1. we're not at the beginning of the row
+          // 2. But: if the current item is wider than the remaining space after
+          //    accounting for empty columns, reset cursor to start of the row
+          //    rather than causing an empty row.
+          if (widths[idx] > unitsRow - cursor) {
+            cursor = 0;
+          }
+          unitsRowRemaining = 0;
+        }
+        addStartClass = unitsRowRemaining >= widths[idx] || cursor > 0;
+      }
+
+      if (cursor > 0 && cursor + widths[idx] > unitsRow) {
+        // adding the current item would overflow the row, needs start class
+        addStartClass = true;
+        cursor = 0;
+      }
+
+      if (addStartClass) {
+        addClassToEl(idx, `g-start-${breakName}-${cursor + 1}`);
+      }
+
+      addClassToEl(idx, `g-col-${breakName}-${unitsThisItem}`);
+      updateCursor(unitsThisItem, false);
+    }
+  }
+}
+
+/**
+ * Given an array and a length, return a new array of the specified length
+ * containing the elements of the original array, recycled as necessary.
+ *
+ * @param arr The original array.
+ * @param len The desired length of the new array.
+ * @returns The new array.
+ */
+function recycleToLength<T>(arr: T[], len: number): T[] {
```

Comment:
This and `isNA()` could, and probably should, be in `../_utils.ts` but I left them here initially to call out that this file has no external dependencies. 

(Technically I'm using `index.ts` to call `customElements.define()` but that could just as easily be in this file to make it completely standalone.)